### ISSUE TITLE
[IOS] Adjusting the sharingOptions

### ISSUE THREAD

**Issue by leoek** [PSPDFKit/react-native#204]
# Details

This [Code in the wrapper](https://github.com/leoek/react-native-pspdfkit/blob/305b4d0fa2dcac76f6d01eea3761c564f6af36f5/ios/RCTPSPDFKit/Converters/RCTConvert%2BPSPDFConfiguration.m#L374) looks like it's possible to adjust the sharing options from react-native. However it is not clear to me how that configuration works.

I guess I have to set the property `sharingConfigurations` in my configuration object, but it is not clear what the value of that property should be.

My current goal is to disable any possibility to print or share the document on ios.

# Acceptance Criteria

- [ ] Add an example for `sharingConfigurations` to [Readme](https://github.com/PSPDFKit/react-native/blob/master/README.md) which disables any possibillity to share or print the document.
- [ ] Add an example for `sharingConfigurations` to [Readme](https://github.com/PSPDFKit/react-native/blob/master/README.md) which declares all or some of the sharing options. The goal should be that it is clear how to set the `sharingConfigurations` properly.

# Additional Information

My current ios configuration:
```
[OS.IOS]: {
        document: path,
        configuration: {
          documentLabelEnabled: false,
          allowToolbarTitleChange: false,
          shouldAskForAnnotationUsername: false,
          createAnnotationMenuEnabled: false,
          editableAnnotationTypes: resolvedAnnotationTypes
        }
      }
```

**Comment by radazzouz** [comment]
Hi @leoek,

`sharingConfigurations` allows you to configure the sharing options. For usage, please refer to the "Custom Sharing Options" example from the Catalog app here: https://github.com/PSPDFKit/react-native/blob/master/samples/Catalog/Catalog.ios.js#L195

> My current goal is to disable any possibility to print or share the document on ios.

If you're looking to disable sharing and printing is to remove those button from the toolbar. You can accomplish this by customize the toolbar in native code (Objective-C). See our blog post for the full implementation details: https://pspdfkit.com/blog/2018/how-to-extend-react-native-api/#customizing-the-ui-using-native-code

Here's how to customize the toolbar on iOS, in Objective-C: https://pspdfkit.com/guides/ios/current/customizing-the-interface/customizing-the-toolbar/

We also have a working example in in this repo which shows how to do this on https://github.com/PSPDFKit/react-native/tree/customize-the-toolbar-in-native-code

In your case, you need to make sure that `activityButtonItem` and `printButtonItem` are excluded from  your `PSPDFViewController`'s navigation item's right bar button items array.

It would look very similar to 

https://github.com/PSPDFKit/react-native/blob/customize-the-toolbar-in-native-code/ios/RCTPSPDFKit/RCTPSPDFKitViewManager.m#L37

In Objective-C, it would look like this:

```objc
// Default toolbar without the print and the share buttons.
[view.pdfController.navigationItem setRightBarButtonItems:@[view.pdfController.thumbnailsButtonItem, view.pdfController.outlineButtonItem, view.pdfController.searchButtonItem, view.pdfController.annotationButtonItem] animated:NO];
```

I am marking this issue as solved, but please do not hesitate to open a new request if you have any other questions.

Thank you!

**Comment by leoek** [comment]
@radazzouz 
Being able to adjust which buttons are shown in the toolbar seems like an essential feature. Also I would rather have this in the core library, than in a fork to avoid merging trouble in the future.

Also I need this feature on ios and android, while hacking this into the objective-c code would only cover ios.

**Comment by radazzouz** [comment]
@leoek,

This is technically feasible, but there’s more to it than just exposing the button. We'd need to expose the native button object and its properties, along with all the related callbacks and delegates. It is way easier to do it in native code.

If you're looking to do the same on Android, we have an example in the same branch ([customize-the-toolbar-in-native-code](https://github.com/PSPDFKit/react-native/tree/customize-the-toolbar-in-native-code)) in https://github.com/PSPDFKit/react-native/blob/customize-the-toolbar-in-native-code/android/src/main/java/com/pspdfkit/views/toolbar/CustomGrouping.java

**Comment by leoek** [comment]
@radazzouz 

It would be enough if there was a flag for the sharing button and one for the print button, which would hide that button. It wouldn't be necessary to expose all the properties and callbacks.

Maybe this could be blacklist (or whitelist) for the buttons in the toolbar.

**Comment by radazzouz** [comment]
Thanks for the suggestion, @leoek!

I can see how blacklisting or whitelisting the buttons in the toolbar would help you with your use case. However, this will only solve the visibility use case. We need to also consider uses cases such as the ability to reorder buttons and the ability to add custom buttons if we were to offer full toolbar customization in the Javascript APIs. We also need to make sure that the same APIs are consistent on iOS, Android and Windows.

For now, I would recommend using the solution which involves minimal alteration of the native code, but I am re-opening this ticket for visibility, as we always welcome pull request contributions from the community.

**Comment by leoek** [comment]
@radazzouz This is already implemented for windows. https://github.com/PSPDFKit/react-native/blob/6215cea58df41e18cf6a4b053dd3aadfe59e0a50/samples/Catalog/Catalog.windows.js#L444

Visibility of the buttons seems much more important to me, than adding new ones or reordering them.

**Comment by leoek** [comment]
@radazzouz 

Could you add such an option for IOS?

**Comment by radazzouz** [comment]
Hi @leoek,

Unfortunately adding such an option is no trivial task. So this can't be done quickly in a few commits. That's because on iOS, unlike on on Windows, we have two toolbars: The navigation toolbar (https://pspdfkit.com/guides/ios/current/customizing-the-interface/customizing-the-toolbar/) and the annotation toolbar (https://pspdfkit.com/guides/ios/current/customizing-the-interface/customize-the-annotation-toolbar/). So the API that you mentioned will not work easily for iOS.

We strive for API consistency across all platforms and this is something we definitely want to improve upon in the future. At the moment, we don't know how this API should look. This feature requires more thought. There are many things that need to be taken into account, such as the iOS navigation toolbar's left and right bar buttons, the fact that we can't have the same button on the left and the right, the order of the buttons, etc. We also need to investigate how to offer a similar API for the annotation toolbar, which is iOS-specific.

I'll bring it up to the rest of the team for further discussion. In the meantime, I suggest proceeding with the workaround we discussed - offering that as React Native API will be a longer project

Thank you!

**Comment by leoek** [comment]
@radazzouz I guess we don't have another option at the moment.

Something I noticed:
Even if we fork the react-native wrapper we cannot adjust the configuration, because `PspdfConfiguration` is part of the closed source library. If the wrapper is supposed to be build upon, this is really unfortunate.

**Comment by radazzouz** [comment]
> Even if we fork the react-native wrapper we cannot adjust the configuration, because PspdfConfiguration is part of the closed source library.

Is this related to this issue or is it a separate question regarding updating `PSPDFConfiguration` in Objective-C in a fork? 

If you're looking to update or set additional `PSPDFConfiguration` properties in a fork in Objective-C, you can do so in https://github.com/PSPDFKit/react-native/blob/master/ios/RCTPSPDFKit/RCTPSPDFKitViewManager.m#L39, like so:

```objc
RCT_CUSTOM_VIEW_PROPERTY(configuration, PSPDFConfiguration, RCTPSPDFKitView) {
  if (json) {
    [view.pdfController updateConfigurationWithBuilder:^(PSPDFConfigurationBuilder *builder) {
      [builder setupFromJSON:json];
      // Set additional configuration properties here;
    }];
  }
}
```

**Comment by leoek** [comment]
I wanted to have a boolean property in the configuration which disabled any sharing options. However `PSPDFConfiguration.h` is part of the ios lib, which is not part of this wrapper. Therefore we cannot simply add any properties there.

**Comment by radazzouz** [comment]
We're working on adding Javascript API customize the toolbar buttons in #233.

**Comment by leoek** [comment]
@radazzouz That sound great! We are currently using your suggestion https://github.com/PSPDFKit/react-native/issues/204#issuecomment-479550507 as a workaround.

**Comment by tiagobnobrega** [comment]
Hi, I just tried the  [#204 (comment)](https://github.com/PSPDFKit/react-native/issues/204#issuecomment-495298848) workaround with no luck. Nothing seems to change the behaviour of and display of the buttons. 

I also tried using the complete code from [RCTPSPDFKitViewManager.m](https://github.com/PSPDFKit/react-native/blob/customize-the-toolbar-in-native-code/ios/RCTPSPDFKit/RCTPSPDFKitViewManager.m#L37) and [RCTPSPDFKitView.m](//github.com/PSPDFKit/react-native/blob/customize-the-toolbar-in-native-code/ios/RCTPSPDFKit/RCTPSPDFKitView.m). In a brand new project. 

I don't have Native objective C/Swift experience. Is there something more to it ? Invalidating a cache or calling some build manually ? Or just change the code, save and run **react-native run-ios** ?

**Comment by radazzouz** [comment]
@tiagobnobrega if you're looking to customize the toolbar buttons in Objective-C, please refer to https://github.com/PSPDFKit/react-native/issues/204#issuecomment-479550507 instead.

Otherwise, you can try and do it Javascript API from the branch on https://github.com/PSPDFKit/react-native/pull/233

Please reach out to our support portal at pspdfkit.com/support/request/ if you any other questions.

**Comment by tiagobnobrega** [comment]
Ok... I figured it out. I'm using it like this at the React component:
```
PSPDFKit.present(filePath, CONFIGURATION);
```
So Obviously changing the  **RCTPSPDFKitViewManager.m** and **RCTPSPDFKitView.m** had no effect.
I had to change the **RCTPSPDFKitManager.m** like this:
```
RCT_EXPORT_METHOD(present:(PSPDFDocument *)document withConfiguration:(PSPDFConfiguration *)configuration) {
  PSPDFViewController *pdfViewController = [[PSPDFViewController alloc] initWithDocument:document configuration:configuration];

//=== Added to set toolbars buttons:
  [pdfViewController.navigationItem setRightBarButtonItems: @[pdfViewController.thumbnailsButtonItem, pdfViewController.outlineButtonItem, pdfViewController.searchButtonItem, pdfViewController.annotationButtonItem] animated:NO];
//===
  
  UINavigationController *navigationController = [[UINavigationController alloc] initWithRootViewController:pdfViewController];

  UIViewController *presentingViewController = RCTPresentedViewController();
  [presentingViewController presentViewController:navigationController animated:YES completion:nil];
}
```

**Comment by tiagobnobrega** [comment]
Thx @radazzouz for the support. I really appreciate it. The features from branch on #223 will be a great addition to the library

**Comment by Andrea-Arguello** [comment]
I am currently using the 1.24.0 version. I would like to disable the sharing option, and I am using the Native Module like @tiagobnobrega. I have tried his solution, but it doesn't seem to have any effect on the toolbar. How do I manage to disable only this button?

**Comment by Andrea-Arguello** [comment]
I am currently working on a project on tag 1.24.0. Since I am also using PSPDFKit.present(), @tiagobnobrega's solution worked for me on iOS; however, since my app is used on both Android and iOS, I wanted to know if it was possible to edit the code natively to achieve the same solution. I have tried commenting the enableShare and enablePrinting on the ConfigurationAdapter.java. However, this doesn't seem to do anything, as the sharing button is still present in the toolbar.

TL;DR: How do I implement @tiagobnobrega's solution for Android? Leaving configureShowShareAction and configureShowPrintAction from ConfigurationAdapter.java with only configuration.disableShare() and  configuration.disablePrinting() doesn't hide the buttons.

**Comment by tiagobnobrega** [comment]
@Andrea-Arguelo. For Android, the react-native options worked. I might be able to share my config here on monday.

**Comment by Andrea-Arguello** [comment]
@tiagobnobrega I managed to hide the share button altogether by changing the build inside ConfigurationAdapter.java
```diff
public PdfActivityConfiguration build() {
+      configuration.disableShare();
+      configuration.disablePrinting();
        return configuration.build();
    }
```
I'd still like to see your config for the react-native options, since that seems like a simpler solution, instead of changing the source files for both iOS and Android on each build.

**Comment by tiagobnobrega** [comment]
Here it is (I'm disabling a bunch of other stuff):
```javascript
const CONFIGURATION = {
  scrollContinuously: false,
  showPageNumberOverlay: true,
  pageScrollDirection: 'horizontal',
  showShareAction: false,
  documentLabelEnabled: false,
  showPrintAction: false,
  createAnnotationMenuEnabled: false,
  annotationGroupingEnabled: false,
  naturalDrawingAnnotationEnabled: false,
  showAnnotationMenuAfterCreation: false,
  shouldAskForAnnotationUsername: false,
  annotationEntersEditModeAfterSecondTapEnabled: false,
  enableAnnotationEditing: false,
  allowToolbarTitleChange: false,
  autosaveEnabled: false,
  hideNavigationBar: true
};
...
//and then called like:
 PSPDFKit.present(filePath, CONFIGURATION);

```

**Comment by Andrea-Arguello** [comment]
@tiagobnobrega thank you so much!
